libraries/Microsoft.Bot.Connector/Authentication/AdalAuthenticator.cs

```diff
@@ -0,0 +1,188 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using Microsoft.IdentityModel.Clients.ActiveDirectory;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Net.Http.Headers;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Bot.Connector.Authentication
+{
+    public class ThrottleException : Exception
+    {
+        public RetryParams RetryParams { get; set; }
+    }
+    
+    public class AdalAuthenticator 
+    {
+        // Our ADAL context. Acquires tokens and manages token caching for us.
+        private readonly AuthenticationContext authContext;
+
+        // Semaphore to control concurrency while refreshing tokens from ADAL.
+        // Whenever a token expires, we want only one request to retrieve a token. 
+        // Cached requests take less than 0.1 millisecond to resolve, so the semaphore doesn't hurt performance under load tests
+        // unless we have more than 10,000 requests per second, but in that case other things would break first.
+        private static Semaphore tokenRefreshSemaphore = new Semaphore(1, 1);
+        private static readonly TimeSpan SemaphoreTimeout = TimeSpan.FromSeconds(10);
+
+        // Depending on the responses we get from the service, we update a shared retry policy with the RetryAfter header
+        // from the HTTP 429 we receive.
+        // When everything seems to be OK, this retry policy will be empty.
+        // The reason for this is that if a request gets throttled, even if we wait to retry that, another thread will try again right away.
+        // With the shared retry policy, if a request gets throttled, we know that other threads have to wait as well.
+        // This variable is guarded by the authContextSemaphore semphore. Don't modify it outside of the semaphore scope.
+        private static volatile RetryParams currentRetryPolicy;
```

Comment:
Given that it's read/written protected by a Sempahore - which presumably introduces a read fence, I suspect the volatile here isn't needed.

Reply:
As I said in another comment, I would suggest not trying to hand roll this yourself for these kinds of reasons. I would lean on the investments that have been made in solving these problems by the community and pull in Polly.

Reply:
Thanks both! I didn't know about Polly, looked into it and loved it! Great pointer.

I will *definitely* make this use polly. The only question is whether we want to push this now and then move to polly in 4.3 (Feb 14th) or delay this and change it to polly (I'm quite booked so the delay is just not implementation time, includes re-load test everything, etc.)

If you see the current code before my change instead of using adal, it hand-writes the adal internal behavior. This proved problematic and error prone, and there are several customers blocked by a bug on that code. This code has been working reliably in V3 for a while. We need to decide whether we want to push this which will unblock customers and has been tested in large V3 bots without issues. I will bring it up in the ship room so the team discusses.

Regardless of the timing, I will get this to Polly for 4.3, definitely great idea.

Reply:
There are a few more features that we want to add for 4.3 also which are 

1) allow ADAL to receive the httpclient from bot options. This will unblock some customers that are behind a proxy
2) Allow ADAL to receive a custom implementation of the token cache. This is awesome for large bots that scale horizontally. For example they could have their token cache be redis based for example so when users hit another instance they still don't have to hit the AAD api but get the token from the cache.